byzcoin/contract_deferred.go

```diff
@@ -0,0 +1,405 @@
+	// This method should do the following:
+	//   1. Parse the input buffer
+	//   2. Compute and store the instruction hashes
+	//   3. Save the data
+	//
+	// Spawn should have those input arguments:
+	//   - proposedTransaction ClientTransaction
+	//   - expireBlockIndex uint64
+	cout = coins
+
+	// Find the darcID for this instance.
+	var darcID darc.ID
+	_, _, _, darcID, err = rst.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
+	}
+
+	// 1. Reads and parses the input
+	proposedTransaction := ClientTransaction{}
+	err = protobuf.Decode(inst.Spawn.Args.Search("proposedTransaction"), &proposedTransaction)
+	if err != nil {
+		return nil, nil, errors.New("couldn't decode proposedTransaction: " + err.Error())
+	}
+	expireBlockIndex := binary.LittleEndian.Uint64(inst.Spawn.Args.Search("expireBlockIndex"))
+	if err != nil {
+		return nil, nil, errors.New("couldn't convert expireBlockIndex: " + err.Error())
+	}
+	numExecution := defaultNumExecution
+
+	// 2. Computes the hashes of each instruction and store it
+	hash := make([][]byte, len(proposedTransaction.Instructions))
+	for i, proposedInstruction := range proposedTransaction.Instructions {
+		hash[i] = hashDeferred(proposedInstruction, inst.InstanceID.Slice())
+	}
+
+	// 3. Saves the data
+	data := DeferredData{
+		ProposedTransaction: proposedTransaction,
+		ExpireBlockIndex:    expireBlockIndex,
+		InstructionHashes:   hash,
+		NumExecution:        numExecution,
+	}
+	var dataBuf []byte
+	dataBuf, err = protobuf.Encode(&data)
+	if err != nil {
+		return nil, nil, errors.New("couldn't encode DeferredData: " + err.Error())
+	}
+
+	sc = append(sc, NewStateChange(Create, inst.DeriveID(""),
+		ContractDeferredID, dataBuf, darcID))
+	return
+}
+
+func (c *contractDeferred) Invoke(rst ReadOnlyStateTrie, inst Instruction, coins []Coin) (sc []StateChange, cout []Coin, err error) {
+	// This method should do the following:
+	//   - Handle the "addProof" invocation
+	//   - Handle the "execProposedTx" invocation
+	//
+	// Invoke:addProof should have the following input argument:
+	//   - identity darc.Identity
+	//   - signature []byte
+	//   - index uint32 (index of the instruction wrt the transaction)
+	err = c.checkInvoke(rst, inst.Invoke)
+	if err != nil {
+		return nil, nil, errors.New("checks of invoke failed: " + err.Error())
+	}
+
+	cout = coins
+
+	// Find the darcID for this instance.
+	var darcID darc.ID
+
+	_, _, _, darcID, err = rst.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
+	}
+
+	switch inst.Invoke.Command {
+	case "addProof":
+		// This invocation appends the identity and the corresponding signature,
+		// which is based on the stored instruction hash (in instructionHashes)
+
+		// Get the given index
+		indexBuf := inst.Invoke.Args.Search("index")
+		if indexBuf == nil {
+			return nil, nil, errors.New("index args is nil")
+		}
+		index := binary.LittleEndian.Uint32(indexBuf)
```

Comment:
We are working with a transaction containing multiple instructions.
The index is necessary because each instruction must be signed independently (ie. they each require a different signature and can potentially have a different DARC).